src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java

```diff
@@ -345,85 +346,19 @@ default MemorySegment allocate(long byteSize) {
-     * @throws IllegalArgumentException if {@code arenaSize <= 0}.
-     * @throws IllegalStateException if {@code session} is not {@linkplain MemorySession#isAlive() alive}.
-     * @throws WrongThreadException if this method is called from a thread other than the thread
-     * {@linkplain MemorySession#ownerThread() owning} {@code session}.
-     */
-    static SegmentAllocator newNativeArena(long arenaSize, MemorySession session) {
-        return newNativeArena(arenaSize, arenaSize, session);
-    }
-
-    /**
-     * Creates an arena-based allocator used to allocate native segments. The returned allocator features
-     * the given block size {@code B} and the given arena size {@code A}, and the native segments
-     * it allocates are associated with the provided memory session.
+     * Returns a segment allocator which responds to allocation requests by returning consecutive slices
+     * obtained from the provided segment. Each new allocation request will return a new slice starting at the
+     * current offset (modulo additional padding to satisfy alignment constraint), with given size.
      * <p>
-     * The allocator arena is first initialized by {@linkplain MemorySession#allocate(long) allocating} a
-     * native segment {@code S} of size {@code B}. The allocator then responds to allocation requests in one of the following ways:
-     * <ul>
-     *     <li>if the size of the allocation requests is smaller than the size of {@code S}, and {@code S} has a <em>free</em>
-     *     slice {@code S'} which fits that allocation request, return that {@code S'}.
-     *     <li>if the size of the allocation requests is smaller than the size of {@code S}, and {@code S} has no <em>free</em>
-     *     slices which fits that allocation request, allocate a new segment {@code S'}, with size {@code B},
-     *     and set {@code S = S'}; the allocator then tries to respond to the same allocation request again.
-     *     <li>if the size of the allocation requests is bigger than the size of {@code S}, allocate a new segment {@code S'},
-     *     which has a sufficient size to satisfy the allocation request, and return {@code S'}.
-     * </ul>
-     * <p>
-     * This segment allocator can be useful when clients want to perform multiple allocation requests while avoiding the
-     * cost associated with allocating a new off-heap memory region upon each allocation request.
-     * <p>
-     * The returned allocator might throw an {@link OutOfMemoryError} if the total memory allocated with this allocator
-     * exceeds the arena size {@code A}, or the system capacity. Furthermore, the returned allocator is not thread safe.
-     * Concurrent allocation needs to be guarded with synchronization primitives.
+     * When the returned allocator cannot satisfy an allocation request, e.g. because a slice of the provided
+     * segment with the requested size cannot be found, an {@link IndexOutOfBoundsException} is thrown.
```

Comment:
This type of exception appears a bit strange to me.

Reply:
Given that the implementation of this method is based on `MemorySegment::asSlice`, I think the exception type is not too bad.